packages/ember-data/lib/system/model/internal-model.js

```diff
@@ -0,0 +1,696 @@
+    model._relationships[key] = createRelationshipFor(model, descriptor, model.store);
+  });
+};
+
+InternalModel.prototype = {
+  isEmpty: retrieveFromCurrentState('isEmpty'),
+  isLoading: retrieveFromCurrentState('isLoading'),
+  isLoaded: retrieveFromCurrentState('isLoaded'),
+  isDirty: retrieveFromCurrentState('isDirty'),
+  isSaving: retrieveFromCurrentState('isSaving'),
+  isDeleted: retrieveFromCurrentState('isDeleted'),
+  isNew: retrieveFromCurrentState('isNew'),
+  isValid: retrieveFromCurrentState('isValid'),
+  dirtyType: retrieveFromCurrentState('dirtyType'),
+
+  constructor: InternalModel,
+  materializeRecord: function() {
+    // lookupFactory should really return an object that creates
+    // instances with the injections applied
+    this.record = this.type._create({
+      id: this.id,
+      store: this.store,
+      container: this.container
+    });
+    this.record._internalModel = this;
+    this._triggerDeferredTriggers();
+  },
+
+  recordObjectWillDestroy: function() {
+    this.record = null;
+  },
+
+
+  deleteRecord: function() {
+    this.send('deleteRecord');
+  },
+
+  save: function() {
+    var promiseLabel = "DS: Model#save " + this;
+    var resolver = Ember.RSVP.defer(promiseLabel);
+
+    this.store.scheduleSave(this, resolver);
+    return resolver.promise;
+  },
+
+  startedReloading: function() {
+    this.isReloading = true;
+    if (this.record) {
+      set(this.record, 'isReloading', true);
+    }
+  },
+
+  finishedReloading: function() {
+    this.isReloading = false;
+    if (this.record) {
+      set(this.record, 'isReloading', false);
+    }
+  },
+
+  reload: function() {
+    this.startedReloading();
+    var record = this;
+    var promiseLabel = "DS: Model#reload of " + this;
+    return new Promise(function(resolve) {
+      record.send('reloadRecord', resolve);
+    }, promiseLabel).then(function() {
+      record.didCleanError();
+      return record;
+    }, function(reason) {
+      record.didError();
+      throw reason;
+    }, "DS: Model#reload complete, update flags")['finally'](function () {
+      record.finishedReloading();
+      record.updateRecordArrays();
+    });
+  },
+
+  getRecord: function() {
+    if (!this.record) {
+      this.materializeRecord();
+    }
+    return this.record;
+  },
+
+  unloadRecord: function() {
+    this.send('unloadRecord');
+  },
+
+  eachRelationship: function(callback, binding) {
+    return this.type.eachRelationship(callback, binding);
+  },
+
+  eachAttribute: function(callback, binding) {
+    return this.type.eachAttribute(callback, binding);
+  },
+
+  inverseFor: function(key) {
+    return this.type.inverseFor(key);
+  },
+
+  setupData: function(data) {
+    var changedKeys = mergeAndReturnChangedKeys(this._data, data);
+    this.pushedData();
+    if (this.record) {
+      this.record._notifyProperties(changedKeys);
+    }
+    this.didInitalizeData();
+  },
+
+  becameReady: function() {
+    var self = this;
```

Comment:
Self is a real thing in the bowser / webWorkers. we should likely not shadow it. Instead lets call this `internalModel`

Or even better, use the non-closure retaining form of schedule

```
Ember.run.schedule('actions', this, this.store.recordArrayManager.recordWasLoaded, this)
```